### PR TITLE
HTTP/2 for Jetty connector

### PR DESCRIPTION
**Pull request by senivam**

### CODE REVIEW

File: connectors/jetty-connector/src/main/java/org/glassfish/jersey/jetty/connector/JettyConnector.java [hunk 1]
```diff
@@ -224,6 +220,37 @@ class JettyConnector implements Connector {
...
+
+    /**
+     * provides custom registered {@link HttpClient} if any (or NULL)
+     *
+     * @param config configuration where {@link HttpClient} could be registered
+     * @return {@link HttpClient} instance if any was previously registered or NULL
+     *
+     * @since 2.40
+     */
+    public HttpClient getRegisteredHttpClient(Configuration config) {
```
Comment (jansupol) [inline connectors/jetty-connector/src/main/java/org/glassfish/jersey/jetty/connector/JettyConnector.java]: Why is this public?

File: connectors/jetty-connector/src/main/java/org/glassfish/jersey/jetty/connector/JettyConnector.java [hunk 2]
```diff
@@ -142,21 +144,15 @@ class JettyConnector implements Connector {
      * @param jaxrsClient JAX-RS client instance, for which the connector is created.
      * @param config client configuration.
      */
-    JettyConnector(final Client jaxrsClient, final Configuration config) {
+    public JettyConnector(final Client jaxrsClient, final Configuration config) {
```
Comment (jansupol) [inline connectors/jetty-connector/src/main/java/org/glassfish/jersey/jetty/connector/JettyConnector.java]: protected

File: connectors/jetty-connector/src/main/java/org/glassfish/jersey/jetty/connector/JettyConnector.java (hunk 1 above)
Comment (jansupol) [reply connectors/jetty-connector/src/main/java/org/glassfish/jersey/jetty/connector/JettyConnector.java]: protected

File: connectors/jetty-http2-connector/src/main/java/org/glassfish/jersey/jetty/http2/connector/JettyHttp2Connector.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    /**
+     * provides required {@link HttpClientTransport} for client
+     *
+     * The overriden method provides {@link HttpClientTransportOverHTTP2} with initialized {@link HTTP2Client}
+     *
+     * @return {@link HttpClientTransportOverHTTP2}
+     * @since 2.40
+     */
+    @Override
+    public HttpClientTransport initClientTransport() {
```
Comment (jansupol) [inline connectors/jetty-http2-connector/src/main/java/org/glassfish/jersey/jetty/http2/connector/JettyHttp2Connector.java]: protected

File: connectors/jetty-http2-connector/src/main/java/org/glassfish/jersey/jetty/http2/connector/JettyHttp2Connector.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    /**
+     * provides custom registered {@link HttpClient} (if any) with HTTP/2 support
+     *
+     * @param config configuration where {@link HttpClient} could be registered
+     * @return {@link HttpClient} instance if any was previously registered or NULL
+     *
+     * @since 2.40
+     */
+    @Override
+    public HttpClient getRegisteredHttpClient(Configuration config) {
```
Comment (jansupol) [inline connectors/jetty-http2-connector/src/main/java/org/glassfish/jersey/jetty/http2/connector/JettyHttp2Connector.java]: protected

File: connectors/jetty-http2-connector/src/main/java/org/glassfish/jersey/jetty/http2/connector/JettyHttp2ClientSupplier.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+import org.eclipse.jetty.http2.client.HTTP2Client;
+import org.eclipse.jetty.http2.client.http.HttpClientTransportOverHTTP2;
+import org.glassfish.jersey.jetty.connector.JettyConnector;
+import org.glassfish.jersey.jetty.connector.JettyHttpClientContract;
+import org.glassfish.jersey.jetty.connector.JettyHttpClientSupplier;
+
+/**
+ * HTTP/2 enabled version of the {@link JettyHttpClientSupplier}
+ *
+ * @since 2.40
```
Comment (jansupol) [inline connectors/jetty-http2-connector/src/main/java/org/glassfish/jersey/jetty/http2/connector/JettyHttp2ClientSupplier.java]: 2.41

File: connectors/jetty-http2-connector/src/main/java/org/glassfish/jersey/jetty/http2/connector/JettyHttp2Connector.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+import org.eclipse.jetty.http2.client.HTTP2Client;
+import org.eclipse.jetty.http2.client.http.HttpClientTransportOverHTTP2;
+import org.glassfish.jersey.jetty.connector.JettyConnector;
+
+import java.util.Optional;
+
+/**
+ * Extends {@link JettyConnector} with HTTP/2 transport support
+ *
+ * @since 2.40
```
Comment (jansupol) [inline connectors/jetty-http2-connector/src/main/java/org/glassfish/jersey/jetty/http2/connector/JettyHttp2Connector.java]: 2.41

File: connectors/jetty-http2-connector/src/main/java/org/glassfish/jersey/jetty/http2/connector/JettyHttp2Connector.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+        return new HttpClientTransportOverHTTP2(new HTTP2Client());
+    }
+
+    /**
+     * provides custom registered {@link HttpClient} (if any) with HTTP/2 support
+     *
+     * @param config configuration where {@link HttpClient} could be registered
+     * @return {@link HttpClient} instance if any was previously registered or NULL
+     *
+     * @since 2.40
```
Comment (jansupol) [inline connectors/jetty-http2-connector/src/main/java/org/glassfish/jersey/jetty/http2/connector/JettyHttp2Connector.java]: 2.41

File: connectors/jetty-http2-connector/src/main/java/org/glassfish/jersey/jetty/http2/connector/JettyHttp2Connector.java [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+        super(jaxrsClient, config);
+    }
+
+    /**
+     * provides required {@link HttpClientTransport} for client
+     *
+     * The overriden method provides {@link HttpClientTransportOverHTTP2} with initialized {@link HTTP2Client}
+     *
+     * @return {@link HttpClientTransportOverHTTP2}
+     * @since 2.40
```
Comment (jansupol) [inline connectors/jetty-http2-connector/src/main/java/org/glassfish/jersey/jetty/http2/connector/JettyHttp2Connector.java]: 2.41

File: connectors/jetty-connector/src/main/java/org/glassfish/jersey/jetty/connector/JettyConnector.java [hunk 9]
```diff
@@ -224,6 +220,37 @@ class JettyConnector implements Connector {
...
+        return new HttpClientTransportOverHTTP();
+    }
+
+    /**
+     * provides custom registered {@link HttpClient} if any (or NULL)
+     *
+     * @param config configuration where {@link HttpClient} could be registered
+     * @return {@link HttpClient} instance if any was previously registered or NULL
+     *
+     * @since 2.40
```
Comment (jansupol) [inline connectors/jetty-connector/src/main/java/org/glassfish/jersey/jetty/connector/JettyConnector.java]: 2.41

File: connectors/jetty-connector/src/main/java/org/glassfish/jersey/jetty/connector/JettyConnector.java [hunk 10]
```diff
@@ -224,6 +220,37 @@ class JettyConnector implements Connector {
         this.cookieStore = client.getCookieStore();
     }
 
+    /**
+     * provides required HTTP client transport for client
+     *
+     * the default transport is {@link HttpClientTransportOverHTTP}
+     *
+     * @return instance of {@link HttpClientTransport}
+     * @since 2.40
```
Comment (jansupol) [inline connectors/jetty-connector/src/main/java/org/glassfish/jersey/jetty/connector/JettyConnector.java]: 2.41

File: connectors/jetty-connector/src/main/java/org/glassfish/jersey/jetty/connector/JettyConnector.java [hunk 11]
```diff
@@ -224,6 +220,37 @@ class JettyConnector implements Connector {
...
+     */
+    public HttpClientTransport initClientTransport() {
```
Comment (jansupol) [inline connectors/jetty-connector/src/main/java/org/glassfish/jersey/jetty/connector/JettyConnector.java]: protected

**Review by jansupol — CHANGES_REQUESTED** [review]

**Review by jansupol — APPROVED** [review]

**Review by jbescos — APPROVED** [review]

**Pull request merged**